File: music-service/src/main/java/com/muco/musicservice/interfaces/dto/ErrorResponseDTO.java
package com.muco.musicservice.interfaces.dto;

import com.muco.musicservice.global.enums.ErrorCode;
import org.springframework.validation.BindingResult;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public record ErrorResponseDTO(
        int status,
        String code,
        String message,
        List<FieldErrorResponseDTO> fieldErrors
) {
    private ErrorResponseDTO(ErrorCode errorCode) {
        this(errorCode.getStatus(), errorCode.getCode(), errorCode.getMessage(), new ArrayList<>());
    }

    private ErrorResponseDTO(ErrorCode errorCode, String message) {
        this(errorCode.getStatus(), errorCode.getCode(), message, new ArrayList<>());
    }

    private ErrorResponseDTO(ErrorCode errorCode, List<FieldErrorResponseDTO> fieldErrors) {
        this(errorCode.getStatus(), errorCode.getCode(), errorCode.getMessage(), fieldErrors);
    }

    public static ErrorResponseDTO of(ErrorCode errorCode) {
        return new ErrorResponseDTO(errorCode);
    }

    public static ErrorResponseDTO of(ErrorCode errorCode, String message) {
        return new ErrorResponseDTO(errorCode, message);
    }

    public static ErrorResponseDTO of(ErrorCode errorCode, BindingResult result) {
        return new ErrorResponseDTO(errorCode, FieldErrorResponseDTO.of(result));
    }

    private record FieldErrorResponseDTO(
            String field,
            String value,
            String reason
    ) {
        private static List<FieldErrorResponseDTO> of(String field, String value, String reason) {
            return new ArrayList<>() {{
                add(new FieldErrorResponseDTO(field, value, reason));
            }};
        }

        private static List<FieldErrorResponseDTO> of(BindingResult result) {
            return result.getFieldErrors().stream()
                    .map(fieldError -> new FieldErrorResponseDTO(
                            fieldError.getField(),
                            fieldError.getRejectedValue() != null ? fieldError.getRejectedValue().toString() : "",
                            fieldError.getDefaultMessage()
                    ))
                    .collect(Collectors.toList());
        }
    }
}

File: chat-service/src/main/kotlin/com/muco/chatservice/domain/interfaces/dto/ErrorResponseDTO.kt
package com.muco.chatservice.domain.interfaces.dto

import com.muco.chatservice.global.enums.ErrorCode
import org.springframework.validation.BindingResult
import reactor.core.publisher.Mono

data class ErrorResponseDTO(
    val status: Int,
    val code: String,
    val message: String,
    val fieldErrors: List<FieldErrorResponseDTO>
) {
    private constructor(errorCode: ErrorCode): this(status = errorCode.status, code = errorCode.code, message = errorCode.message, fieldErrors = emptyList())
    private constructor(errorCode: ErrorCode, message: String): this(status = errorCode.status, code = errorCode.code, message, fieldErrors = emptyList())
    private constructor(errorCode: ErrorCode, fieldErrors: List<FieldErrorResponseDTO>): this(status = errorCode.status, code = errorCode.code, message = errorCode.message, fieldErrors)

    companion object {
        fun of(errorCode: ErrorCode): Mono<ErrorResponseDTO> = Mono.just(ErrorResponseDTO(errorCode))
        fun of(errorCode: ErrorCode, message: String): Mono<ErrorResponseDTO> = Mono.just(ErrorResponseDTO(errorCode, message))
        fun of(errorCode: ErrorCode, result: BindingResult): Mono<ErrorResponseDTO> = Mono.just(ErrorResponseDTO(errorCode, FieldErrorResponseDTO.of(result)))
    }

    data class FieldErrorResponseDTO(
        val field: String,
        val value: String,
        val reason: String?
    ) {
        companion object {
            fun of(field: String, value: String, reason: String): List<FieldErrorResponseDTO> = listOf(
                FieldErrorResponseDTO(field, value, reason)
            )

            fun of(result: BindingResult): List<FieldErrorResponseDTO> = result.fieldErrors.stream()
                    .map {
                        FieldErrorResponseDTO(
                            field = it.field,
                            value = (it.rejectedValue ?: "").toString(),
                            reason = it.defaultMessage)
                    }
                    .toList()
        }
    }
}
